### PR TITLE
MCR-2822 IIIF Thumbnail also use linked image

### PR DESCRIPTION
**Pull request by eckstueck**
added function to use linked image if there is no derivate

[Link to jira](https://mycore.atlassian.net/browse/MCR-2822).

### CODE REVIEW

File: mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java [hunk 1]
```diff
@@ -66,15 +67,33 @@ protected MCRTileInfo createTileInfo(String id) throws MCRIIIFImageNotFoundExcep
...
+                                maindoc);
+                            if (tileInfoForFile.isPresent()) {
+                                return tileInfoForFile.get();
                             }
                         }
                     }
                 }
             }
+            Optional<String> linkedImage = MCRLinkTableManager.instance()
+                .getDestinationOf(mcrID, MCRLinkTableManager.ENTRY_TYPE_DERIVATE_LINK).stream().findFirst();
```
Comment (sebhofmann) [inline mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: The implementation iterates over all derivates and the first one which has a tiles will be used. 
It should work the same way with derivate_links and not only consider the first link.

File: mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java [hunk 2]
```diff
@@ -66,15 +67,33 @@ protected MCRTileInfo createTileInfo(String id) throws MCRIIIFImageNotFoundExcep
...
+            if (linkedImage.isPresent()) {
+                MCRObjectID deriID = MCRObjectID.getInstance(
+                    linkedImage.get().substring(0, linkedImage.get().indexOf("/")));
+                MCRDerivate deri = MCRMetadataManager.retrieveMCRDerivate(deriID);
+                if (deri.getDerivate().getClassifications().stream()
+                    .map(classi -> classi.getClassId() + ":" + classi.getCategId())
```
Comment (sebhofmann) [inline mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: This can be MCRCategoryID::toString

**Review by sebhofmann — CHANGES_REQUESTED** [review]
Not direct related to the branch but i think `for (String type : derivateTypes) {` in line 62 and `.anyMatch(type::equals)` in line 66 could be replaced with `.anyMatch(derivateTypes::contains)`

File: mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java (hunk 2 above)
Comment (eckstueck) [reply mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: deri.getDerivate().getClassifications() returns a ArrayList<MCRMetaClassification> and MCRMetaClassification::toString does not return the right format.

Comment (rsteph-de) [reply mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: May be we should a the following methods to `MCRMetaClassification`:
```
public final MCRCategoryID getCategoryID(){
        return category;
}

public final void setValue(MCRCategoryID category) throws MCRException {
        this.category = category;
    }
```
In my opinion it is useful to have the MCRCategoryID object available.

Comment (eckstueck) [reply mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: Should I add this in this Pull request?

Comment (rsteph-de) [reply mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: @sebhofmann  and I came to the conclusion to leave it as it is for now

**Review by sebhofmann — APPROVED** [review]

File: mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java [hunk 3]
```diff
@@ -49,40 +51,90 @@ public MCRThumbnailImageImpl(String implName) {
...
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Optional<MCRTileInfo> createTileInfoForFile(String derID, String file) {
+        final MCRTileInfo mcrTileInfo = new MCRTileInfo(derID, file, null);
+        final Optional<Path> tileFile = this.tileFileProvider.getTileFile(mcrTileInfo);
+        if (tileFile.isPresent() && Files.exists(tileFile.get())) {
```
Comment (sebhofmann) [inline mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: can be written as return tileFile.filter(Files::exists);

Comment (eckstueck) [reply mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: I cant, because i need to return the MCRTileInfo and not the Path.

**Review by rsteph-de — APPROVED** [review]
O.K. for me

File: mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java (hunk 3 above)
Comment (sebhofmann) [reply mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRThumbnailImageImpl.java]: Then you could use:
```
       final MCRTileInfo tileInfo = new MCRTileInfo(derID, file, null);
       return Optional.of(tileInfo).filter(t-> this.tileFileProvider.getTileFile(t).filter(Files::exists).isPresent());
```

**Pull request merged**